src/Collections/Sequences/Seq.dfy

```diff
@@ -0,0 +1,720 @@
+  {
+    reveal HasNoDuplicates();
+    var c := a + b;
+    if |c| > 1 {
+      assert forall i, j {:trigger c[i], c[j]}:: i != j && 0 <= i < |a| && |a| <= j < |c| ==>
+        c[i] in multiset(a) && c[j] in multiset(b) && c[i] != c[j]; 
+    }
+  }
+
+  /* A sequence with no duplicates converts to a set of the same cardinality */
+  lemma LemmaCardinalityOfSetNoDuplicates<T>(s: seq<T>)
+    requires HasNoDuplicates(s)
+    ensures |ToSet(s)| == |s|
+  {
+    reveal HasNoDuplicates();
+    reveal ToSet();
+    if |s| == 0 {
+    } else {
+      LemmaCardinalityOfSetNoDuplicates(DropLast(s));
+      assert ToSet(s) == ToSet(DropLast(s)) + {Last(s)};
+    }
+  }
+
+  /* proves that there are no duplicate values in the multiset version of the sequence */
+  lemma LemmaMultisetHasNoDuplicates<T>(s: seq<T>)
+    requires HasNoDuplicates(s)
+    ensures forall x {:trigger multiset(s)[x]} | x in multiset(s):: multiset(s)[x] == 1
+  {
+    if |s| == 0 {
+    } else {
+      assert s == DropLast(s) + [Last(s)];
+      assert Last(s) !in DropLast(s) by {
+        reveal HasNoDuplicates();
+      }
+      assert HasNoDuplicates(DropLast(s)) by {
+        reveal HasNoDuplicates();
+      }
+      LemmaMultisetHasNoDuplicates(DropLast(s));
+    }
+  }
+
+  /* finds the index of a certain element in the sequence */
+  function method {:opaque} IndexOf<T(==)>(s: seq<T>, v: T): (i: nat)
+    requires v in s
+    ensures i < |s| && s[i] == v
+  {
+    if s[|s|-1] == v then |s| - 1 else IndexOf(s[..|s|-1], v)
+  }
+
+  /* finds the index of a certain element in the sequence if found */
+  function method {:opaque} IndexOfOption<T(==)>(s: seq<T>, v: T): (o: Option<nat>)
+    ensures if o.Some? then o.value < |s| && s[o.value] == v else v !in s
+  {
+    if |s| == 0 then None()
+    else if s[|s|-1] == v then Some(|s| - 1) else IndexOfOption(s[..|s|-1], v)
+  }
+
+  /* slices out a specific position's value from the sequence */
+  function method {:opaque} Remove<T>(s: seq<T>, pos: nat): (s': seq<T>)
+    requires pos < |s|
+    ensures |s'| == |s| - 1
+    ensures forall i {:trigger s'[i], s[i]} | 0 <= i < pos :: s'[i] == s[i]
+    ensures forall i {:trigger s'[i]} | pos <= i < |s| - 1 :: s'[i] == s[i+1]
+  {
+    s[..pos] + s[pos+1..]
+  }
+
+  /* slices out a specific value from the sequence */
+  function method {:opaque} RemoveValue<T(==)>(s: seq<T>, v: T): (s': seq<T>)
+    ensures v !in s ==> s == s'
+    ensures v in s ==> |multiset(s')| == |multiset(s)| - 1
+    ensures v in s ==> multiset(s')[v] == multiset(s)[v] - 1
+    ensures HasNoDuplicates(s) ==> HasNoDuplicates(s') && ToSet(s') == ToSet(s) - {v}
+  {
+    reveal HasNoDuplicates();
+    reveal ToSet();
+    if v !in s then s
+    else
+      var i := IndexOf(s, v);
+      assert s == s[..i] + [v] + s[i+1..];
+      s[..i] + s[i+1..]
+  }
+
+  /* slices out a specific value from the sequence if found */
+  function method {:opaque} RemoveValueOption<T(==)>(s: seq<T>, o: Option<T>): (s': seq<T>)
```

Comment:
Hold up, is this function actually useful? I would `if o.Some? then RemoveValue(s, o.Get) else s` would be simpler and fit better into the caller's logic anyway.

Perhaps you were thinking more along the lines of `RemoveValueOption<T(==)>(s: seq<T>, v: T): (s': seq<T>, o: Option<T>)`, along the same lines as `IndexOf` vs. `IndexOfOption` above. Although in that case I'd say just returning a boolean would be better.

Reply:
That makes sense, I removed this function.